### PR TITLE
[AArch64] Add streaming-mode stack hazards.

### PR DESCRIPTION
**Pull request by davemgreen**
Under some SME contexts, a coprocessor with its own separate cache will be used for FPR operations. This can create hazards if the CPU and the SME unit try to access the same area of memory, including if the access is to an area of the stack.

To try to alleviate that, this patch attempts to introduce extra padding into the stack frame between FP and GPR accesses, controlled by the StackHazardSize option. Without changing the layout of the stack frame, a stack object of the right size is added between GPR and FPR CSRs. Another is added to the stack objects section, and stack objects are sorted so that FPR > Hazard padding slot > GPRs (where possible).

Unfortunately some things are not handled well (VLA area, FPR arguments on the stack, object with both GPR and FPR accesses), but if those are controlled by the user then the entire stack frame becomes GPR at the start/end with FPR in the middle, surrounded by Hazard padding. This can greatly help reduce something that can be difficult for the user to control themselves.

The current implementation is opt-in through an -aarch64-stack-hazard-size flag, and should have no effect if the option is unset. In the long run the implementation might change (for example using more base pointers to separate in more cases, re-enabling ldp/stp using an extra register, etc), but this gets at least something for people to use in llvm-19 if they need it.  The only change whilst the option is unset will be a fix for making sure the stack increment is added at the right place when it cannot be converted to postinc (++MBBI). I believe without extra padding that can not normally be reached.

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64FrameLowering.cpp [hunk 1]
```diff
@@ -4595,16 +4768,28 @@ void AArch64FrameLowering::orderFrameObjects(
         GB.AddMember(TaggedFI);
       else
         GB.EndCurrentGroup();
+
+      if (AFI.hasStackHazardSlotIndex()) {
```
Comment (efriedma-quic) [inline llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:4772]: Please rearrange this code so it isn't mixed with the stack tagging code.

File: llvm/lib/Target/AArch64/AArch64FrameLowering.cpp [hunk 2]
```diff
@@ -4537,9 +4703,15 @@ bool FrameObjectCompare(const FrameObject &A, const FrameObject &B) {
   //
   // If all else equal, sort by the object index to keep the objects in the
   // original order.
-  return std::make_tuple(!A.IsValid, A.ObjectFirst, A.GroupFirst, A.GroupIndex,
+  if (A.IsValid != B.IsValid)
+    return A.IsValid;
+  if (A.Accesses == 2 && B.Accesses != 2)
+    return true;
+  if (A.Accesses == 8 && B.Accesses != 2)
+    return true;
```
Comment (efriedma-quic) [inline llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:4711]: Please rewrite this to something that's more obviously a strict weak ordering.

File: llvm/lib/Target/AArch64/AArch64FrameLowering.cpp [hunk 3]
```diff
@@ -3798,6 +3952,10 @@ bool AArch64FrameLowering::enableStackSlotScavenging(
   // function doesn't use a FP.
   if (AFI->hasStreamingModeChanges() && !hasFP(MF))
     return false;
+  // Don't allow register salvaging with hazard slots, in case it moves objects
+  // into the wrong place.
+  if (AFI->hasStackHazardSlotIndex())
+    return false;
```
Comment (efriedma-quic) [inline llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:3958]: What's the interaction here?  Just that scavenging could introduce hazards?

**Review by efriedma-quic — COMMENTED** [review]
A couple ways to possibly save some stack space:

- For FP CSRs, would it make sense to go through integer registers?  e.g. `fmov x16, d8; str x16, [sp, #16]`.  More microops, but you don't need hazard padding for the FP CSRs.  Maybe depends on how expensive the micro-ops are.
- If the hazard is related to cache-lines, realigning the stack could reduce the amount of necessary padding.

**Review by davemgreen — COMMENTED** [review]
> A couple ways to possibly save some stack space:
> 
> - For FP CSRs, would it make sense to go through integer registers? e.g. fmov x16, d8; str x16, [sp, 16]. More microops, but you don't need hazard padding for the FP CSRs. Maybe depends on how expensive the micro-ops are.
> - If the hazard is related to cache-lines, realigning the stack could reduce the amount of necessary padding.

Thanks - we were hoping to get something into the LLVM-19 compiler so the amount of changes was deliberately kept minimal. In the longer term we might try and change more. Realigning the stack (or portions of it) is certainly an option.

If the SME unit is separate from the CPU and the SME unit hold the values of vector registers, the fmov d->gpr might be a relatively expensive operation. Depending on the function, we might need to stack frames on top of one-another and with the current scheme we end up with `GPR (CPU) => GPU > hazard > FPR > FPR > hazard > GPR => GPR > hazard > FPR > ...`, so I think in the general case you need two hazard paddings. If you don't have one of the regions (or any sub-calls) then things might become simpler though.

**Review by efriedma-quic — APPROVED** [review]
LGTM

File: llvm/test/CodeGen/AArch64/stack-hazard.ll [hunk 4]
```diff
@@ -0,0 +1,2812 @@
...
+; CHECK1024-NEXT:    .cfi_restore b12
+; CHECK1024-NEXT:    .cfi_restore b13
+; CHECK1024-NEXT:    .cfi_restore b14
+; CHECK1024-NEXT:    .cfi_restore b15
+; CHECK1024-NEXT:    ret
+entry:
+  tail call void @other()
+  ret i32 %x
+}
+declare void @other()
```
Comment (kmclaughlin-arm) [inline llvm/test/CodeGen/AArch64/stack-hazard.ll:2812]: Is it worth also adding one test which is locally-streaming?

**Review by kmclaughlin-arm — APPROVED** [review]
Thanks @davemgreen, I've just let one minor comment on the test file but otherwise this LGTM.

**Comment by davemgreen** [comment]
Thanks for the reviews.

**Pull request merged**